### ISSUE TITLE
Openshift SCC should be installed before any resource which creates a pod

### ISSUE THREAD

**Issue by ashokponkumar** [helm/helm#8439]
When creating a [SCC](https://docs.openshift.com/enterprise/3.0/admin_guide/manage_scc.html) and service account as part of a helm chart, since it is not ordered to be created before deployment and other resources, the scc does not have an effect on the pods. 

The simplest fix might be to just add kind "SecurityContextConstraints" just after "ServiceAccount" in the kind_sorter.go file.

https://github.com/helm/helm/blob/release-3.0/pkg/releaseutil/kind_sorter.go

Output of `helm version`:
version.BuildInfo{Version:"v3.1.2", GitCommit:"d878d4d45863e42fd5cff6743294a11d28a9abce", GitTreeState:"clean", GoVersion:"go1.13.8"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-26T06:16:15Z", GoVersion:"go1.14", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.2+18cfcc9", GitCommit:"18cfcc9", GitTreeState:"clean", BuildDate:"2020-05-25T09:09:52Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
ROKS (RedHat Openshift Kubernetes service - IBM Cloud)

**Comment by hickeyma** [comment]
@ashokponkumar The kind sorter is for Kubernetes resources only. Is "SecurityContextConstraints" an Open Shift object?

**Comment by ashokponkumar** [comment]
@hickeyma I am not sure of the difference between a kubernetes resource and an Openshift object. Isn't everything in Kubernetes resource, even if it is added as a CRD. Each flavour/version of kubernetes has its own set of kinds depending on the operators that are installed in it.

For example, `kubectl api-resources` does lists `SecurityContextConstraints` along with every other resource supported in that cluster.

**Comment by liuming-dev** [comment]
cc #8292 , and this need reviewers.

**Comment by liuming-dev** [comment]
whether this can be solved via using hook?

**Comment by ashokponkumar** [comment]
hooks don't solve this issue for couple of reasons:

1. Conceptually, hooks are things that need to be done outside the chart context. SCC's and many other CRs are part of the chart.
2. If a scc is created as part of the chart, then it should be deleted when the chart is deleted. Hooks don't get deleted as far as I understand.

Technically speaking, the right way to handle any CR would be by using the "weight" concept that is used in the helm-hooks. That way it would be very flexible. 

@liuming-dev  I looked at #8292, and if I read it right, it is using alphabetical ordering or ordering in the manifest. I would assume both are of no significance when it comes to Kind names. Isn't it? Why can't we use weight annotation similar to that used by hooks?

**Comment by mattfarina** [comment]
@ashokponkumar There is a difference between conformant Kubernetes and Kubernetes with customizations. Kubernetes conformance is something handled by the CNCF and something that can install to a conformant Kubernetes should be able to be installed into any conformant Kubernetes system.

`SecurityContextConstraints` is an OpenShift addition that's not part of the conformance.

Helm sorting targets conformant Kubernetes. If the sorting targeted every possible CRD variation it would be a fair amount of work to just handle the public ones. There are many private and proprietary ones that Helm could not handle because it does not have the option of knowing about them.

The best way would be for Kubernetes to hold an ordering but the Kubernetes project considers things to be declarative so that order should not matter. I would not expect to get one from the Kubernetes project.

So, how can Helm handle this in a manner that does not hard code CRD ordering and is backwards compatible in v3? This is an honest question.

**Comment by ashokponkumar** [comment]
Thanks @mattfarina. I completely agree with you. Just for my own understanding, I had been trying to understand the Kubernetes flavors, and it would really help me, if you can point to the components that are part of the conformant Kubernetes distribution.

If I were to propose a generic solution that covers all CRs, then it would go as follows:

1. Have an annotation by the name "helm.sh/weight", which takes an Integer value (-ve or +ve). This a concept that Helm anyway uses for hooks ("helm.sh/hook-weight").
2. We retain the InstallOrder. We install CRs which have a negative weight in ascending order before those in InstallOrder. Then those with InstallOrder, and then Install those with +ve weights in ascending order.
3. If anyone specifies a weight for a resource in Install order, which is +ve or -ve, it goes with that batch. 
4. If the weight is 0, then it happens just before those in the install order. So if there is a resource which in InstallOrder, but has a weight of 0, it will be applied before any other resource.

@mattfarina What do you think about this proposal?

**Comment by hickeyma** [comment]
@ashokponkumar Would it be possible to use a separate chart for the SCC and deploy it prior to the app chart?

**Comment by ashokponkumar** [comment]
> @ashokponkumar Would it be possible to use a separate chart for the SCC and deploy it prior to the app chart?

Technically yes, but it is not ideal for two reasons.

1. If I have 3 such resources which are dependent on order, then I would need to use 3 charts, which is not scalable.
2. I would always love to have all the resources in a single folder for a single application for context, and this would make it a bit more messy.

**Comment by paulczar** [comment]
Regardless of Helm's ability to determine when to apply an `SCC` I would have some concerns about putting them in the same chart as deployment. Clearly I am not you and don't know your full context and reasoning ... but these are my thoughts on wrapping stuff like `SCC` or `PSP` up in a helm chart.

Reading the documentation for Security Context Constraints, it seems to me they're designed to be set by the cluster administration level, not the application deployer. Setting `SCC` as part of a Helm install is like giving someone `sudo` access to only run a script they control. Ideally `RBAC` would be set so that a regular app deployer doesn't even have access to create/modify `SCC`.

I would second @hickeyma that `SCC` should be done in a separate chart and run as part of preparing a cluster for applications to be deployed. If a given Chart requires special `SCC` then that should be handled seperately, otherwise you're effectively allowing an application deployer to set an `SCC` that allows pretty much anything.

As an example of how I solve similar, I have a set of resources that I always deploy as the cluster admin, things like prometheus, fluentd, cert-manager, etc, as well as all my default Pod Security Policies etc,   These are all separate charts that I wrap up in a [helmfile](https://github.com/roboll/helmfile) chart.   If anyone requires a resource that is blocked by my `PSP` I decide if they should be able to do it and update my charts/helmfile appropriately at the cluster level.

**Comment by ashokponkumar** [comment]
> Regardless of Helm's ability to determine when to apply an `SCC` I would have some concerns about putting them in the same chart as deployment. Clearly I am not you and don't know your full context and reasoning ... but these are my thoughts on wrapping stuff like `SCC` or `PSP` up in a helm chart.

Valid points. Just to give some more context, I am trying to create a helm chart based operator. The servicename used by the operator generally runs with more permissions than required by the application. So I am trying to let the helm chart create a scc that is limited in permission, and use it for its resources. When I have to add a serviceName to a scc, I have to edit the scc. So instead of polluting other scc's for each of my deployment, I am creating a new scc with only the relevant servicename for each deployment. This is more to limit the permissions I give to my resources/pods. 

> If I were to propose a generic solution that covers all CRs, then it would go as follows:
> 
> 1. Have an annotation by the name "helm.sh/weight", which takes an Integer value (-ve or +ve). This a concept that Helm anyway uses for hooks ("helm.sh/hook-weight").
> 2. We retain the InstallOrder. We install CRs which have a negative weight in ascending order before those in InstallOrder. Then those with InstallOrder, and then Install those with +ve weights in ascending order.
> 3. If anyone specifies a weight for a resource in Install order, which is +ve or -ve, it goes with that batch.
> 4. If the weight is 0, then it happens just before those in the install order. So if there is a resource which in InstallOrder, but has a weight of 0, it will be applied before any other resource.

Do you think it makes sense to use this as a generic CR implementation, so that it can handle such resources, which are dependent on others? Users can fall back to the dual chart approach based on their need and usage pattern. Its gives a choice.

**Comment by paulczar** [comment]
@ashokponkumar there have been some attempts in the past (example https://github.com/helm/helm/pull/5214) to implement chart template ordering, but they haven't been followed through to a mergeable state.  I would suggest we'd be open to the idea, but I don't see it as a roadmapped item that the maintainers are actively working on.

**Comment by ashokponkumar** [comment]
> @ashokponkumar there have been some attempts in the past (example #5214) to implement chart template ordering, but they haven't been followed through to a mergeable state. I would suggest we'd be open to the idea, but I don't see it as a roadmapped item that the maintainers are actively working on.

Yes. I too noticed it. Example #3635 seems pretty close to this suggestion. I did not see any objection to this idea though. If the commenters in this thread see this as an useful idea, I will create a new issue for this one.

**Comment by ashokponkumar** [comment]
Closing this in favor of #8446, since it covers a more generic solution to this issue.

**Comment by hickeyma** [comment]
Re-opened as it contains comments so far. Updated as proposal.

**Comment by hickeyma** [comment]
@ashokponkumar This proposal will be discussed at [Helm dev weekly meeting](https://github.com/helm/helm#community-discussion-contribution-and-support) on this Thursday 16th. It would be great if you could attend to pitch the proposal.

**Comment by ashokponkumar** [comment]
sure @hickeyma . Will attend.

**Comment by liuming-dev** [comment]
> @ashokponkumar This proposal will be discussed at the [Helm dev weekly meeting](https://github.com/helm/helm#community-discussion-contribution-and-support) on this Thursday, the 16th. It would be great if you could attend to pitch the proposal.

PR #8292 is the same topic,  I wish this PR could be discussed together.  :pray:

**Comment by bacongobbler** [comment]
There are multiple threads all asking for similar requests. A short list of those threads:

- #8291
- #7072
- #6851
- #5916
- #5642

I would propose closing all of these threads in favour of #6851. I'm not sure why there's a need to track discussions for each individual resource here.

**Comment by ashokponkumar** [comment]
> There are multiple threads all asking for similar requests. A short list of those threads:
> 
> * #8291
> * #7072
> * #6851
> * #5916
> * #5642
> 
> I would propose closing all of these threads in favour of #6851. I'm not sure why there's a need to track discussions for each individual resource here.

@bacongobbler Though the initial discussion started with an individual resource, we converged on a generic solution for handling all CRs using a weight concept as seen in the new modified description. Also we have a PR #8448 attached to this issue which has this fix implemented.

Repasting the updated description for ease of reference:

```
Similar to how we have an ordering requirement for known resources in the InstallOrder, users want the capability to order custom resources too. For example a custom resource like SCC need to be applied before creation of any pods.

To enable this capability we can follow the same approach we use for hooks. Essentially add a annotation "helm.sh/order-weight" which can take an integer value (+ve or -ve). This will behave the same way as "helm.sh/hook-weight", but for resources not covered by hooks.

Here is how it will work:

The first resources to be installed will be all resources that have negative weights, and then those with zero or no weights, and then those with positive weights. All in ascending order.
For resources which have same weight, Install order is preserved for known resources.

PR here showing the solution.
```

**Comment by ashokponkumar** [comment]
To address the question raised in today's developer call around dependent charts, one way to handle it could be to have a chart weight specifiable along with the dependency specification in chart.yaml. The value specified in it will get added as a "helm.io/chart-weight" annotation in each of those charts resources, when the resources are templated. In case of dependency within a dependency, the "helm.io/chart-weight" will be the sum of the all chart-weights encountered till now.

And while each resource order-weight is computed, the chart-weight gets added to its value. For the base chart, the value is 0.

**Comment by jecnua** [comment]
Issue still valid, please remove stale label.

**Comment by jecnua** [comment]
Issue still valid, please remove stale label. (I really need to get back to my project of an anti-stale-bot bot for github).

**Comment by henrikb123** [comment]
I have a similar problems as addressed here. I want a CR in my chart to be installed before the general application (deployment, service, ...). I even want it to go before a job with helm hook-annotation also part of the chart, which further complicates things.

A suggested solution that would solve my problem and possibly cover other cases discussed here:

Make it possible for resources with hook-annotations to become managed objects. For example I want my CR with hook-weight -2 to be managed, but my job with hook-weight -1 to be un-managed as currently supported.

If there was an annotation like
helm.sh/hook-object-ownership: managed

then I would be able to update my CR object using Helm and also avoid the deletion of the object that comes with hooks (before-hook-creation, hook-succeeded, hook-failed) and is unwanted for my CR object.

**Comment by ashb** [comment]
Still not stale. Can one of the helm committers pin this issue please 😃

**Comment by jecnua** [comment]
Still not stale, doing a little dance for the third not stale comment:
┏(･o･)┛♪┗ (･o･) ┓

**Comment by pampatzoglou** [comment]
bump

**Comment by janecekt** [comment]
Hi all - I desperately need this functionality. I need to make sure that my custom resources are installed in the right order. The way I see it there are two ways how to do this:

1) Add support "hook.sh/weight" annotation to each manifest - manifests will be sorted first by weight and then by kind (as per the proposal described above).

2) Add support for customizing of the KindSortOrder in the chart. The user would define in Chart.yaml how the custom resources should be interleaved with standard resources. This would be compiled into customized KindSortOrder which would then be used by the kind_sorter in a usual way. NOTE: this could be done such that only new kinds can be added but the order of existing kinds would not modified.

These approaches are not mutually exclusive, so they can both be implemented simultaneously. I am happy to implement this change + the necessary tests etc. (it should not take more than one weekend).

I understand from the discussions that this is somewhat controvertial feature. Since, I'd like my change to be integrated into the mainline - would it be possible for the maintainers of helm to advise which approach they prefer? Is there a consenus of the maintainers if/how this should be done? What would be the best way to arrive at a design that could be accepted and integrated into the mainline.

**Comment by bergerx** [comment]
@janecekt If you want to drive this, seems like the best path forward to push for this is putting together a [HIP](https://github.com/helm/community/blob/main/hips/hip-0001.md) with the best path forward that you can think off, collect comments and keep improving HIP as suggestions come in.

Once an agreement is achieved in the HIP, kicking off the implementation and getting it merged into the mainstream repo should be straightforward. But seeking blessing to start working on some implementation in GitHub issues like this one is recently seems to be not fruitful.

**Comment by adrafiq** [comment]
Bump. Need this!

**Comment by wenzizone** [comment]
any updates on this?  this is really useful

**Comment by liqlin2015** [comment]
Yea. We have the same problem.
We have the operator component deployed as helm chart and set operator CR in chart template. Helm can help create the operator CR after the other kubernetes resources, that's good. But for uninstallation, we need the operator CR to be cleaned up before anything, the reverse order of installation. 
Looking forward a solution for this.

**Comment by obesser** [comment]
What is currently blocking it? It is obviously a highly requested feature and there were good suggestions to solve it that came up here. In addition, it seems it can be implemented as a non-breaking change (correct me if I'm wrong).

**Comment by omerurhan** [comment]
We are still waiting for this feature. Any updates?

**Comment by myjaffoo** [comment]
I have been exploring using Helm and the Pulumi Kubernetes Operator together to orchestrate the deployment of infrastructure and application within a single Helm chart but experienced this issue of Helm not knowing when the Pulumi Stack custom resource was "ready" i.e. all cloud resources were created successfully, before moving on to deploying my workload which had a dependency on the cloud resource.

Therefore I created a "not so elegant" but functional workaround application called the [custom resource status checker](https://github.com/appvia/cr-status-checker) which is deployed as a Kubernetes Job within the same helm chart and leveraged the Helm hooks to orchestrate the following ordering of execution i.e. 

1. Pulumi Stack custom resource due to no hook present
2. Custom resource status checker due to the addition of the `post-install` hook with a negative hook-weight of `-5`
3. Workload / cloud resource consumer due to the addition of the `post-install` hook with a positive hook-weight of `5`

This solution works if Helm is then ran with the `--wait` and `--wait-for-jobs` flags e.g.
```bash
helm upgrade --install some-cloud-consumer ./some-cloud-consumer-chart --wait --wait-for-jobs
```

N.B - You can influence the Kubernetes Job Controller's retry logic by setting the `.spec.backoffLimit`.

**Comment by withnale** [comment]
Any update on this one?

**Comment by pawellegowski89** [comment]
The lack of this functionality appears in many new tickets and PR on Helm. This is a big problem for custom resources (with custom kind). Not CRD, only CR for example [https://github.com/bitnami-labs/sealed-secrets](url) You can find dozens of questions online: **When will it be finally completed? When Helm allow sorting object with custom kind during install/upgrade?**
Its existence makes helm not a good package manager when we have CR in our chart.

The current workaround is to split the definitions into two separate charts. This complicates the continuous delivery process very much. Helm pre-install and pre-upgrade hooks are not working for CR (custom resources). CDR chart dir not helping in this case and using subcharts too. 

Please solve this problem or give us good workaround for this.

@technosophos @bacongobbler

**Comment by alekseysavitskiy** [comment]
managed to sort custom resources using List wrapper:
```
apiVersion: v1
kind: List
items:
- kind: CR1
- kind: CR2
```

**Comment by alekseysavitskiy** [comment]
Solution which I mention does not cover all cases, but it allow to manage order of installing custom resources (CR) by helm. In my case I have 2 CR, one depends on other, but helm sorts and applies manifests in wrong order. Using kind List allows me manually sort order of applied manifests.

**Comment by antaloala** [comment]
Thanks @alekseysavitskiy. I was not sure to follow you so I decided to google a little around that "kind: List" and I finally found this: [https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/list.yaml](https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/list.yaml)  ... and so I now understood the trick ;-)

Yes, it not solves all cases ... but solves some of them ;-)

Thanks a lot for sharing !

**Comment by antaloala** [comment]
@andrejusc kind is not enough to univocally identify a custom resurce; it is also needed the api-group.
Your proposal could then be enhanced using the ``<kind>.<api-group>`` notation

**Comment by andrejusc** [comment]
@antaloala - agreed. We could extend that into overall GVK (i.e. group/version/kind) notation. Mostly seeking here some community support, so we could bring this to whatever steering board behind Helm and make it done sooner than later.

**Comment by gonzaloserrano** [comment]
The `List` hack worked for me too. I still have to try the `"helm.sh/hook-weight"` annotation.

**Comment by motabass** [comment]
Please!!! implement this. We are having the problems with Istio CRs in our helm chart. Istio uses different CRs which should all be installed before any deployments (ServiceMeshMember, ServiceEntry, VirtualService, Gateway). So for now helm is completely unusable for istio/service-mesh.

**Comment by seboss666** [comment]
Hello,
Great to see that beeing sorted out, but not sure if it can solve my current problem : I want to move some ingress configuration to a dedicated file, but when applying, I get thrown an error because it wants to create the new ingress rule with a domain/path that already is another rule (the one I want to split).

I did not find a way in the documentation to enforce the order of ressources of the same kind, may I miss something ?

**Comment by adrafiq** [comment]
@seboss666 that sounds like ingress validating hook error not a helm error. Can you paste the log output here.

**Comment by seboss666** [comment]
No need, we dug a little further and this is an edge case with no definitive solution : it depends on the resource's name, as they are applied in alphabetic order when computed. When we roll back, we can get the error the other way, for the same reason.

So we need to manually act on this specific case by removing the domain before applying this change. And yes, this is a validation problem, Helm definitely cannot help with that. Good thing is that we don't do this every day :D

**Comment by abdennour** [comment]
This is an anti-pattern in k8s. You can reach that by setting write probes

**Comment by abdennour** [comment]
the only way as per today is to delegate helm to argocd & use sync-wave annotations there.

**Comment by andrejusc** [comment]
so on original CRDs flexible sorting aspect support for deployment needs via Helm - is there agreement/disagreement in Helm community? As this issue by now open for 2+ years. Implementation of such support doesn't look to be complex, but is that then affects any other pattern to be aware about?

**Comment by joejulian** [comment]
Happy to have a contribution to address this. Probably should start with a [HIP](https://github.com/helm/community/blob/main/hips/hip-0001.md) also take a look at the [contributing](https://github.com/helm/helm/blob/main/CONTRIBUTING.md) doc.

**Comment by pawellegowski89** [comment]
Solution is here: 
https://github.com/helm/helm/pull/9534
https://github.com/helm/community/pull/230

@NiklasWagner please fix this PR according to the reviewer's comments. 
We are waiting for this solution!

**Comment by sergey-kudriavtsev** [comment]
### Hi guys!
I have an alternative suggestion..
1. I am a diagram developer, so I understand exactly in what order they should be installed.
2. Offering to simply create an **flow** folder, where all the contents, level 1, will be based on alphabetical sorting.
flow/
   1-example.yaml
   ./alpha crd
   ./betta-crd

**Comment by aladdin-atypon** [comment]
This feature should be intuitively provided long ago! I have ExternalSecret objects and these are used by statefulset, the statefulset can't be ready without these secrets created by ExternalSecret.

It's not like we need to create separate chart for such case!

**Comment by Dave-c-Ross** [comment]
I faced another use case today, in a unique chart, I have openshift CRDs MachineSet ClusterAutoscaler and MachineAutoscaler. MachineSet has to be created before MachineAutoscaler, since MachineAutoscaler reference MachineSet.

So I had to create different charts for each kind

**Comment by aladdin-atypon** [comment]
No progress or update?

**Comment by motabass** [comment]
Still using multiple Helm-Charts only to get ordering right. This approach seems so wrong because the idea seemed to be 1 chart for 1 application...now i have to use 2 just to have some CRs being in place before Kubernetes-Kinds. Any plans on solving this? Because all efforts already made seem to be stale...

**Comment by corinz** [comment]
My use case for this feature would be to only update Istio DestinationRule and VirtualService _after_ any new replicasets/pods are ready. Today, they are created together, before pods are ready (even when using `--wait`). Hooks are undesirable because these resources are orphaned/distinct from the release.

Does anyone have a suggestion for a workaround? Could we not allow the sort order slice to be merged with a user input?

**Comment by karmops** [comment]
Regarding this, maybe it could be solved using [(kustomize)](https://www.innoq.com/en/blog/2022/07/advanced-kustomize-features/) instead of helm.

**Comment by owais** [comment]
I have another unrelated use case custom ordering could help with. I implemented a post-renderer for our internal usage. The post-renderer is quite important and we cannot afford a person or system not using it as it can result in very significant diffs that can be destructive. We also need to provide some additional context to the post-renderer. Usually this can be done via post renderers args but the data we need to provide is contextual and usually is stored in different values files for different environments. We solved both of these issues by adding a custom template that an unsupported resource called "HelmContext". The resource uses data from values to render itself. The post renderer program extracts information it needs from this helm context resource and operates on the rest of the resources based on this info. The post renderer program removes this helm context resource from the output so it is never installed in k8s. This gives us a very nice side-effect i.e, the release install would fail without going through the post renderer implicitly enforcing using post-renderer. 

Now the issue we face is that custom resources are applied after everything else so this works only when CI runs `helm diff` or `helm template` before `install`. Someone manaully running `install` would still see the failure but other resources would be applied before the error would be thrown. 

Being able to tell helm to always install our custom resource before anything else would very elegantly solve this issue for us :)

**Comment by hajdukda** [comment]
Just hit the same issue with Custom Resources for Vault secrets :).. Custom Resource should be installed first coz it's a dependency for everything else.

**Comment by brc** [comment]
:popcorn: :smiley:

**Comment by scottrigby** [comment]
See "H4HIP: Helm Sequencing Proposal" https://github.com/helm/community/pull/373

**Comment by pawellegowski89** [comment]
What is the status of this error?
What are we actually waiting for?

The entire community has been struggling with this for at least four years. Please fix it already, because how long can we wait and find workarounds!

**Comment by TerryHowe** [comment]
> What is the status of this error? What are we actually waiting for?

My interpretation is we are waiting  for a contributor to step up and implement it. A design has been approved, so that is a big first step on a complex problem.

**Comment by helltraitor** [comment]
Headache on migration job, vault (crd) and deployment. I see solution in special annotations:

```yaml
someResource:
  annotations:
    helm.sh/need: 'ConfigMap:v1:*-specificName,Secret:v1:*' # all secrets needed
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
```

`helm.sh/need: Kind:Version:RegexpName[,Kind:Version:RegexpName]*`

Something like this (no matter how exactly it's looking, the point is: Helm will just build dependency tree and execute in order required to user. No weights and magic numbers

imho, I don't see reason to use numbers and weights (same logic as package managers in unix)
